lib/etcd.py

```diff
@@ -60,33 +75,89 @@ def fetch_and_install(self, source, sha):
+            return leader_get('cluster')
+
+    def cluster_data(self, unit=None):
+        ''' Non duplicating merge of dictionaries. Each new node adds a key to
+            the unitdata dict. The dict is the persisted state of the cluster
+            to be used if the leader "goes away". Its the only liferaft in the
+            event of an emergency.
+        '''
+        # ...I'm only mostly certain of this
 
-        return cluster.rstrip(',').lstrip(',')
+        cluster_data = self.db.get('etcd.cluster_data')
+
+        if not unit:
+            unit = {self.unit_name: {'private_address': self.private_address,
+                                     'public_address': self.public_address,
+                                     'port': self.port,
+                                     'management_port': self.management_port,
+                                     'unit_name': self.unit_name}
+                    }
+        # De-dupe the data in the databag
+        for k in unit:
+            cluster_data[k] = unit[k]
+        self.db.set('etcd.cluster_data', cluster_data)
+        return cluster_data
 
     def register(self, cluster_data):
+        ''' Perform self registration against the Etcd leader.
+
+            @params cluster_data - a dict of the data to pass to the
+            etcd leader unit to declare the units intent to join the cluster.
+        '''
         if not self.db.get('registered'):
-            command = "etcdctl -C http://{}:4001 member add {}" \
-                      " http://{}:7001".format(cluster_data['leader_address'],
-                                               cluster_data['unit_name'],
-                                               self.private_address)
-            check_call(split(command))
-            self.db.set('registered', True)
+            command = "etcdctl -C http://{}:{} member add {}" \
+                      " http://{}:{}".format(cluster_data['leader_address'],
+                                             self.port,
+                                             cluster_data['unit_name'],
+                                             self.private_address,
+                                             self.management_port)
+            try:
+                check_call(split(command))
+                self.db.set('registered', True)
+            except CalledProcessError:
+                log('Notice: Unit failed to registration command', 'WARNING')
+
+    def unregister(self, cluster_data):
+        # this wont work, it needs to be etcdctl member remove {{GUID}}
+        command = "etcdctl -C http://{}:{} member remove {}" \
+                  " http://{}:{}".format(cluster_data['leader_address'],
+                                         config('port'),
+                                         cluster_data['private_address'],
+                                         config('management_port'))
+        check_call(split(command))
```

Comment:
If this command does not work it will need a similar, try/except as the register() function. I understand there is more work t be done on scale down, but lets not make it throw an error on the way out.